include/aws/crt/iot/DeviceDefender.h

```diff
@@ -0,0 +1,168 @@
+                 */
+                void StopTask() noexcept;
+
+                /**
+                 * Initiates Defender V1 reporting task.
+                 */
+                int StartTask() noexcept;
+
+                /**
+                 * Returns the task status.
+                 */
+                DeviceDefenderV1ReportTaskStatus GetStatus() noexcept;
+
+                OnDefenderV1TaskCancelledHandler OnDefenderV1TaskCancelled;
+
+                void *cancellationUserdata;
+
+                /**
+                 * @return the value of the last aws error encountered by operations on this instance.
+                 */
+                int LastError() const noexcept { return m_lastError; }
+
+              private:
+                Crt::Allocator *m_allocator;
+                DeviceDefenderV1ReportTaskStatus m_status;
+                aws_iotdevice_defender_report_task_config m_taskConfig;
+                aws_iotdevice_defender_v1_task *m_owningTask;
+                int m_lastError;
+
+                DeviceDefenderV1ReportTask(
+                    Crt::Allocator *allocator,
+                    std::shared_ptr<Mqtt::MqttConnection> mqttConnection,
+                    const Crt::String &thingName,
+                    Io::EventLoopGroup &eventLoopGroup,
+                    DeviceDefenderReportFormat reportFormat,
+                    uint64_t taskPeriodNs,
+                    uint64_t networkConnectionSamplePeriodNs,
+                    OnDefenderV1TaskCancelledHandler &&onCancelled = NULL,
+                    void *cancellationUserdata = nullptr) noexcept;
+
+                static void s_onDefenderV1TaskCancelled(void *userData);
+            };
+
+            /**
+             * Represents a builder for creating a DeviceDefenderV1ReportTask object.
+             */
+            class AWS_CRT_CPP_API DeviceDefenderV1ReportTaskBuilder final
+            {
+              public:
+                DeviceDefenderV1ReportTaskBuilder(
+                    Crt::Allocator *allocator,
+                    std::shared_ptr<Mqtt::MqttConnection> mqttConnection,
+                    Io::EventLoopGroup &eventLoopGroup,
+                    const Crt::String &thingName);
+
+                /**
+                 * Sets the device defender report format, or defaults to AWS_IDDRF_JSON.
+                 */
+                DeviceDefenderV1ReportTaskBuilder &WithDeviceDefenderReportFormat(
+                    DeviceDefenderReportFormat reportFormat) noexcept;
+
+                /**
+                 * Sets the task period nanoseconds. Defaults to 5 minutes.
+                 */
+                DeviceDefenderV1ReportTaskBuilder &WithTaskPeriodNs(uint64_t taskPeriodNs) noexcept;
+
+                /**
+                 * Sets the network connection sample period nanoseconds. Defaults to 5 minutes.
+                 */
+                DeviceDefenderV1ReportTaskBuilder &WithNetworkConnectionSamplePeriodNs(
```

Comment:
Why is nano-seconds the sample period unit? Why not something closer to our actual expected max sample rate, say...seconds? Maybe milliseconds, but that would be a lot of sampling.

Reply:
That makes sense. I think seconds as the unit is appropriate. I'll make that change and have it do the conversion to nanos for passing to the C layer.